contracts/chain-adapters/Solana_Adapter.sol

```diff
@@ -0,0 +1,193 @@
+
+    // Custom errors for constructor argument validation.
+    error InvalidCctpTokenMessenger(address tokenMessenger);
+    error InvalidCctpMessageTransmitter(address messageTransmitter);
+
+    // Custom errors for relayMessage validation.
+    error InvalidRelayMessageTarget(address target);
+    error InvalidOriginToken(address originToken);
+    error InvalidDestinationChainId(uint256 destinationChainId);
+
+    // Custom errors for relayTokens validation.
+    error InvalidL1Token(address l1Token);
+    error InvalidL2Token(address l2Token);
+    error InvalidAmount(uint256 amount);
+    error InvalidTokenRecipient(address to);
+
+    /**
+     * @notice Constructs new Adapter.
+     * @param _l1Usdc USDC address on L1.
+     * @param _cctpTokenMessenger TokenMessenger contract to bridge tokens via CCTP.
+     * @param _cctpMessageTransmitter MessageTransmitter contract to bridge messages via CCTP.
+     * @param solanaSpokePool Solana spoke pool address, decoded from Base58 to bytes32.
+     * @param solanaUsdc USDC mint address on Solana, decoded from Base58 to bytes32.
+     * @param solanaSpokePoolUsdcVault USDC token address on Solana for the spoke pool, decoded from Base58 to bytes32.
+     */
+    constructor(
+        IERC20 _l1Usdc,
+        ITokenMessenger _cctpTokenMessenger,
+        IMessageTransmitter _cctpMessageTransmitter,
+        bytes32 solanaSpokePool,
+        bytes32 solanaUsdc,
+        bytes32 solanaSpokePoolUsdcVault
+    ) CircleCCTPAdapter(_l1Usdc, _cctpTokenMessenger, CircleDomainIds.Solana) {
+        // Solana adapter requires CCTP TokenMessenger and MessageTransmitter contracts to be set.
+        if (address(_cctpTokenMessenger) == address(0)) {
+            revert InvalidCctpTokenMessenger(address(_cctpTokenMessenger));
+        }
+        if (address(_cctpMessageTransmitter) == address(0)) {
+            revert InvalidCctpMessageTransmitter(address(_cctpMessageTransmitter));
+        }
+
+        cctpMessageTransmitter = _cctpMessageTransmitter;
+
+        SOLANA_SPOKE_POOL_BYTES32 = solanaSpokePool;
+        SOLANA_SPOKE_POOL_ADDRESS = _trimSolanaAddress(solanaSpokePool);
+
+        SOLANA_USDC_BYTES32 = solanaUsdc;
+        SOLANA_USDC_ADDRESS = _trimSolanaAddress(solanaUsdc);
+
+        SOLANA_SPOKE_POOL_USDC_VAULT = solanaSpokePoolUsdcVault;
+    }
+
+    /**
+     * @notice Send cross-chain message to target on Solana.
+     * @dev Only allows sending messages to the Solana spoke pool.
+     * @param target Program on Solana (translated as EVM address) that will receive message.
+     * @param message Data to send to target.
+     */
+    function relayMessage(address target, bytes calldata message) external payable override {
+        if (target != SOLANA_SPOKE_POOL_ADDRESS) {
+            revert InvalidRelayMessageTarget(target);
+        }
+
+        bytes4 selector = bytes4(message[:4]);
+        if (selector == SpokePoolInterface.setEnableRoute.selector) {
```

Comment:
Why is `setEnableRoute` a specific case here? Is it the only function we'll be calling on the Solana spoke pool? For example, does this prevent us from pausing deposits/fills, emergency deleting root bundles, or forcing through admin root bundles? These haven't really happened. much in the past but the latter two have happened at least once

Reply:
We need to keep the address version of [setEnableRoute](https://github.com/across-protocol/contracts/blob/d97325003a1a38828bb7b87b52be67c3f25ab65a/contracts/SpokePool.sol#L313-L320) for backward compatibility with the hub pool, and that specific handler translates the message to the bytes32 version.

@Reinis-FRP, do you remember why we decided to handle the translation here, which led us to do the casting here instead of handling it in SVM?

Reply:
The main reason is that `setEnableRoute` is invoked via `setDepositRoute` on HubPool that we are not upgrading and we need different types on Solana. For other admin methods they are crafted manually and bridged via generic `relaySpokePoolAdminFunction` method, so the operator can adjust the interface types for all other methods to Solana.

Reply:
But @Reinis-FRP this conversion could also be done in SVM side instead right? (related to https://github.com/across-protocol/contracts/pull/672#discussion_r1831856338)

Reply:
This would be bit tricky as we would need to pass mapping account that stores address translation when receiving the message on Solana side. It does not really belong under `HandleReceiveMessage` which translates all other supported admin calls statelessly. And if we move the address translation account down to self-invoked `SetEnableRoute` then we would end up changing the interface of `set_enable_route` to use `[u8; 20]` as its `origin_token` parameter, which is not ideal as we also support calling it from Solana admin (or need to implement both flavors). And this also requires additional admin method to update the mapping account with correct address translation.

Not saying its impossible to do, but it seemed to have less code complexity if we move the translation to the adapter on EVM side, especially, since it already needs to translate trimmed Solana USDC address to its vault associated token account anyway.

Reply:
So it sounds like the tradeoff here is (less code complexity, special case for `setEnableDepositRoute` on the EVM side, generic SVM side logic) versus (more code complexity, but totally generic EVM logic and bespoke SVM side logic), right? What about gas costs?